compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -0,0 +1,534 @@
+            // u5never as vendor extended type
+            let mut s = String::from("u5never");
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        // Compound types
+        // () in Rust is equivalent to void return type in C
+        ty::Tuple(_) if ty.is_unit() => {
+            typeid.push('v');
+        }
+
+        // Sequence types
+        ty::Tuple(tys) => {
+            // u5tupleI<element-type1..element-typeN>E as vendor extended type
+            let mut s = String::from("u5tupleI");
+            for ty in tys.iter() {
+                s.push_str(&encode_ty(tcx, ty, dict, options));
+            }
+            s.push('E');
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::Array(ty0, len) => {
+            // A<array-length><element-type>
+            let mut s = String::from("A");
+            s.push_str(&format!("{}", &len.val().try_to_scalar().unwrap().to_i64().unwrap()));
+            s.push_str(&encode_ty(tcx, *ty0, dict, options));
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::Slice(ty0) => {
+            // u5sliceI<element-type>E as vendor extended type
+            let mut s = String::from("u5sliceI");
+            s.push_str(&encode_ty(tcx, *ty0, dict, options));
+            s.push('E');
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        // User-defined types
+        ty::Adt(adt_def, substs) => {
+            let mut s = String::new();
+            let def_id = adt_def.0.did;
+            let name = format!("{}", tcx.item_name(def_id));
+            let crate_name = format!("{}", tcx.crate_name(def_id.krate));
+            // FIXME(rcvalle): Identify C char and integer type uses and encode them with their
+            // respective builtin type encoding as specified by the Itanium C++ ABI if
+            // encoding C types is enabled.
+            if name == "c_void"
+                && (crate_name == "core" || crate_name == "std" || crate_name == "libc")
+            {
+                // C char, integer, and void types
+                s.push('v');
+            } else if adt_def.repr().c() {
+                // For for cross-language CFI support, the encoding must be compatible at the FFI
+                // boundary. For instance:
+                //
+                //     struct type1 {};
+                //     void foo(struct type1* bar) {}
+                //
+                // is encoded as:
+                //
+                //     !1 = !{i64 0, !"_ZTSFvP5type1E"}
+                //
+                // So, encode any repr(C) user-defined type [or extern types (i.e., ty::Foreign)] as
+                // <length><name>, where <name> is <unscoped-name>.
+                s.push_str(&format!("{}", name.len()));
+                s.push_str(&name);
```

Comment:
> Hmm, so if I understand correctly, if one would like to use `#[repr(C)]` type for FFI purposes, one would have to guarantee that the type has the same name on Rust side and C/C++ side (and in C++ additionally ensure that the type is not nested in a namespace)?

Yes, they have to use the same name for the type. Also, the Rust compiler provides interoperability with foreign code written in C, not C++. The latter has to be achieved via C wrappers. So, AFAIK, namespaced types can't used at the FFI boundary.

> 
> The `#[repr(C)]` types are often used for their well-known layout, rather than for FFI purposes (e.g., search for `#[repr(C)]` in a tokio repository). Such types can be have generic parameters, and with this approach we lose the ability to distinguish between different generic arguments.

Yes, unfortunately this is a trade off we'll have to make for cross-language CFI support (which is the most important goal).

> 
> This implementation has also cache key issue similar to the one mentioned earlier.

Thanks for pointing it out! I'll fix it.

Reply:
> Yes, unfortunately this is a trade off we'll have to make for cross-language CFI support (which is the most important goal).

For cross-language compatibility that indeed seems inevitable in a way.

It makes me wonder if we should use a different mangling for C calling convention and Rust calling convention, especially that we already have a mangling designed for Rust.

Reply:
As I mentioned in previous comments, I don't think we should be mixing encodings for CFI or using v0's encoding for several reasons but the main ones being separation of concerns (i.e., we shouldn't defer type membership grouping to v0 and its encoding), and losing control of how types are grouped, which will be needed in the future for implementing other features (e.g., increasing or decreasing granularity by adding or removing information, etc.).

Reply:
I don't really follow that argument at all. Neither Itanium mangling nor v0 mangling have anything to say about type membership or grouping. They don't control nor specify any of those things.

Reply:
It's not about the encoding themselves, but about separation of concerns and being able to change and specify the amount and how information is encoded as required for best CFI effectiveness. This encoding's implementation's concern is grouping function types for best CFI effectiveness, while v0's current implementation's concern is mangling symbols for linking.

Reply:
Could you give a few concrete examples? On what basis would you like to group types together?

Reply:
LLVM CFI (i.e., forward-edge control flow protection) aggregates function pointers in groups identified by their return and parameter types, and before indirect calls, test whether the destination pointer is associated with a type identifier (i.e., test type membership) of all possible destinations for that indirect call. For instance:

```rust
fn A(i32) {};
fn B(i32) {};
fn C(i32) {};
```
A type test membership at an indirect call site that calls a function that returns void and receives a single i32 parameter should pass for A, B, and C.

Reply:
Hmm, I still don't follow how this demonstrates  **grouping of types**, since those function have the same type signature so the grouping is trivial?

Reply:
It's just a simple example to explain how LLVM CFI works and how function pointers should be grouped for CFI. What information are you looking for more specifically?

Reply:
I would like to understand why you think we shouldn't use v0 for functions with Rust calling convention. It still seems like very good choice to me, since it designed for Rust and with limited amount of implementation work we could have a first class support in demanglers.

> I don't think we should be mixing encodings for CFI or using v0's encoding for several reasons but the main ones being separation of concerns (i.e., we shouldn't defer type membership grouping to v0 and its encoding), and losing control of how types are grouped, which will be needed in the future for implementing other features (e.g., increasing or decreasing granularity by adding or removing information, etc.).

For example when you mention type membership grouping, I expect the type grouping to be implemented as type level transformation that selects a unique representative type out of an equivalence class of types. It would precede the type mangling process and be completely independent of it. Similarly to control granularity of encoding, we would omit the information from function signature.

I also don't understand what it means to "defer type membership grouping to v0 and its encoding", since v0 mangling doesn't group types.

Reply:
> This is where I think some of the confusion is coming from.These ids are never demangled and aren't expected to be--they're only used by the LowerTypeTests pass to aggregate function pointers by unique ids.

Yes, I understand that. Nevertheless I found the ability to demangle the type identifiers to be invaluable when trying to understand why some of resulting  identifiers are compatible and others aren't.

> For instance, we may want to add information about what registers are clobbered and what are preserved by a function. This can't be done with having control over only a previous type transformation.

Interesting. Indeed this wouldn't fit into this scheme, but ...

> And what about adding granularity? How are we supposed to add information besides type information? Using the example I gave before, how would we add information about what registers are clobbered and what are preserved by a function?

How would it be possible to add more information that isn't completely redundant with respect to information that is already contained in Rust type system? Functions with the same type are supposed to be compatible after all?

Reply:
> How would it be possible to add more information that isn't completely redundant with respect to information that is already contained in Rust type system? Functions with the same type are supposed to be compatible after all?

I gave an example already, but here's another example: Given three functions with the same return and parameter types A, B, and C, and a indirect call at location L to a function with the same return and parameter types as A, B, and C. We may want in the future, If it's determined that it isn't possible for C to be a called at location L, to group C in a different group than A and B. (Although I think it could be done during type transformation by synthesizing types to force these functions into different groups, if possible, this is beyond the type transformation's concern, and may result in ambiguity if not done properly.)

The threat model and threat(s) that control flow protection mitigates are at a lower level than type information, and we shouldn't constrain ourselves to it. Ideally, we should control all aspects of the mitigation, from code generation, type transformation, encoding, to even optimization passes (e.g., for instance, to avoid data loaded into registers that was supposed to be immutable from spilling onto the stack, such as the group ids, and later reused for checks) to avoid introducing new CFI bypasses into the Rust compiler implementation, and to create opportunities to iterate on and improve it later.

Reply:
The motivation for my question was to understand the design choice of Itanium type mangling, instead of v0 mangling for functions with Rust calling convention (where the rationale for compatibility with C/C++ doesn't exists) - so it was essentially question about mangling of type information. If the information is of other nature, I don't see how it motivates the use of one mangling scheme over the other?

Why shouldn't we use the mangling designed for Rust to encode Rust types? Why use a pseudo Itanium mangling instead? In either case we can equally well add other kind of information to the identifiers.

Reply:
It's not about whether to use v0's encoding or not, it's about using current v0's implementation, and more importantly, implementing or using multiple encodings (i.e., mixing encodings). Let me try to summarize:

Why not use current v0's implementation?

* Separation of concerns. v0's currently implementation's concern is mangling symbols for linking, where this encoding's concern is encoding type ids for grouping function pointers for best CFI efficiency. From v0's RFC:
    
      The scheme must provide an unambiguous string encoding for everything that can end up in a binary's symbol table.

  Encoding type ids for grouping function pointers require ambiguity in some cases, and introducing changes to current v0's implementation for that goes against its concern/requirement. (Although I think most it could be done during type transformation, there are the other points about mixing encodings.)

Why not mix encodings?

* Reducing complexity and security risks. For security features, the less complexity the better. Implementing or using multiple encodings may result in unexpected behavior or introducing new CFI bypasses into the Rust compiler implementation.
* Reducing maintenance burden. We'd have to not only maintain feature parity between multiple encodings, but also guarantee that the resulting groupings of function pointers have equivalent properties between them. (Either via type transformation or changing the encodings' implementation for non type information.)

Reply:
> It's not about whether to use v0's encoding or not, it's about using current v0's implementation ...

Thanks for explaining this. I didn't realize that you were referring to the existing implementation of v0 mangling.

> Encoding type ids for grouping function pointers require ambiguity in some cases, and introducing changes to current v0's implementation for that goes against its concern/requirement. (Although I think most of it could be done during type transformation, there are the other points about mixing encodings.)

A short comment about ambiguity and my earlier suggestion to separate type mapping / grouping from type mangling. In the current implementation those two concerns are tightly coupled. During mangling of a type, it might be transformed before mangling continues. The transformation are ambiguous in the sense that during mangling we will encounter distinct types that will eventually mangle the same way. In general that would be fine, but in the presence of a back references it becomes a correctness concern (we need to ensure that the same transformation is applied consistently to types stored / looked up in a type dictionary). Types that mangle the same way have to be treated as fully equivalent at this stage, any ambiguity at this point is a bug. The suggestion to separate those two concerns arisen primarily as a solution to this problem.

Reply:
> A short comment about ambiguity and my earlier suggestion to separate type mapping / grouping from type mangling. In the current implementation those two concerns are tightly coupled. During mangling of a type, it might be transformed before mangling continues. The transformation are ambiguous in the sense that during mangling we will encounter distinct types that will eventually mangle the same way. In general that would be fine, but in the presence of a back references it becomes a correctness concern (we need to ensure that the same transformation is applied consistently to types stored / looked up in a type dictionary). Types that mangle the same way have to be treated as fully equivalent at this stage, any ambiguity at this point is a bug. The suggestion to separate those two concerns arisen primarily as a solution to this problem.

Yes, I thought about two solutions for this substitution dictionary key issue: using a type's encoding string as the key for the dictionary, or doing the type transformation. Both would require traversing a type tree once (either for prepopulating the substitution dictionary, or for transforming types) before traversing it again for encoding. Doing the type transformation is a more elegant solution, and seems the right thing to do. I'll implement it soon.

Reply:
Done.